rfcs/20201121-keras-model-fit-ps.md

```diff
@@ -0,0 +1,501 @@
+
+
+###### Initial focus on `dataset_fn` path
+
+`dataset` path is concurrently explored to provide compatibility with other strategies, although `dataset_fn` path would be the initial focus to make sure we get to the bottom and user experience can be delivered end-to-end. All following sections are based on the aforementioned `dataset_fn` path, and will be updated when the `dataset` path is more understood.
+
+
+###### Proposed design for `dataset` path
+
+This is to be completed.
+
+
+#### The setup of `ClusterCoordinator`
+
+To take advantage of TF2 support of parameter server training, a `ClusterCoordinator` should be created for handling asynchronous function scheduling and joining. The preferred route should be that such an object is abstracted away from the user with `model.fit` training API as an implementation detail.
+
+In terms of who keeps track of the `ClusterCoordinator`, and when it starts allocating threads, there are a few options. Here, we assume that the distribution `Strategy` object can provide whether or not it is supposed to be used with a `ClusterCoordinator`. See below “Changes in tf.distribute” section for more information.
+
+
+##### Option 1: Attach the `ClusterCoordinator`’s lifecycle to `model.fit`
+
+With this option, an attribute is proposed for the `Model` that keeps track of the `ClusterCoordinator`, and it is instantiated at the beginning of `model.fit` time. 
+
+
+```
+class Model(...):
+  def __init__(self):
+    self._cluster_coordinator = None
+    ...
+
+  def fit(self, ...):
+    if hasattr(self.distribute_strategy, 'should_use_with_coordinator'):
+      self._cluster_coordinator = cluster_coordinator.ClusterCoordinator(
+          self.distribute_strategy)
+    ... # the rest of `fit`
+    self._cluster_coordinator.shut_down() # Shut down at the end of `fit`
+    self._cluster_coordinator = None
+
+class ClusterCoodinator(object):
+  def shut_down(self):
+    # Join the threads and terminate resources. We don't have this implemented yet.
+```
+
+
+
+##### Option 2: Have an attribute in `ParameterServerStrategy` that holds the `ClusterCoordinator`
```

Comment:
+1

Reply:
+1

Reply:
+1 as well, the lifetimes of the PSStrategy and Coordinator seem to be the same, whereas the lifetimes of a Model and a Strategy can be different

Reply:
Sounds good. Thanks for the input

Reply:
+1